static_frame/core/frame.py

```diff
@@ -5097,13 +5103,19 @@ def blocks() -> tp.Iterator[np.ndarray]:
         else:
             block_gen = blocks
 
-        if self._columns.depth > 1:
-            raise ErrorInitFrame('cannot unset index with a columns with depth greater than 1')
+        # self._columns._blocks may be None until array cache is updated.
+        self._columns._update_array_cache()
+        column_blocks = self._columns._blocks._blocks
 
         if names:
-            columns = chain(names, self._columns.values)
+            columns = chain(np.array(tuple(names)), column_blocks)
         else:
-            columns = chain(self._index.names, self._columns.values)
+            columns = chain(self._index.names, column_blocks)
```

Comment:
I am not sure this is right.

The blocks obtained from the `IndexHierarchy` are along the axis of the types, which for an `IndexHierarchy` in the columns positions are oriented like "rows":

```python
>>> ih = sf.IndexHierarchy.from_product(('a', 'b'), (1, 2))                                                                                                                                              
>>> ih                                                                                                                                                                                                   
<IndexHierarchy>
a                1
a                2
b                1
b                2
<<U1>            <int64>
>>> ih.values # force a recache call                                                                                                                                                                     
array([['a', 1],
       ['a', 2],
       ['b', 1],
       ['b', 2]], dtype=object)
>>> ih._blocks._blocks                                                                                                                                                                                   
[array(['a', 'a', 'b', 'b'], dtype='<U1'), array([1, 2, 1, 2])]
```
So if the user is providing `names`, and the number of `names` matches the depth of the old index, than we will want to do something like this to concatenate the names in front of the blocks. But here I am duplicating the names for each level of the columns, which is a little strange but I am not sure what else to do.

```python
column_blocks = [concat_resolved(np.array(names), block) for block in blocks]
```
Then you can create the columns with:
```python
columns = self._COLUMNS_HIERARCHY_CONSTRUCTOR._from_type_blocks(TypeBlocks.from_blocks(column_blocks), own_blocks=True)
```

Reply:
Thanks! I was fairly confident that what I was doing wasn't correct. I've updated the code based on your suggestion, and am able to correctly preserve dtypes. 

I had to adapt your suggestion, as `concat_resolved` expects 2d arrays. Here's the result:

```python
        column_blocks = self._columns._blocks._blocks
        names_t = zip(*names)

        column_blocks_new = tuple(
                concat_resolved((np.array([name]), block[np.newaxis]), axis=1).T
                for name, block in zip(names_t, column_blocks)
        )
        column_type_blocks = TypeBlocks.from_blocks(column_blocks_new)
```